C/c4PredictiveQuery.cc

```diff
@@ -40,7 +40,7 @@ class C4PredictiveModelInternal : public PredictiveModel {
         try {
             return _c4Model.prediction(_c4Model.context,
                                        (FLDict)input,
-                                       (c4Database*)dfDelegate,
+                                       dynamic_cast<c4Database*>(dfDelegate),
```

Comment:
What is the distinction, here?  Is dynamic cast checked at runtime?

Reply:
dynamic cast will set the element to null if it does not exceed, so I guess the answer to your question is "yes" ?

Reply:
Oh!  That's interesting!  If it fails the value is null!
Is that what we want here?  What are the circumstances under which that thing isn't a c4Database?

Reply:
I guess we don't expect that to be null here.

Reply:
The reason I added it is that `dynamic_cast<T>` will fail to compile if it hasn't seen the full declaration of `T`. The bug was that `c4Database` had been forward-referenced but not declared in this source file, because it was including `Database.hh` not `c4Database.hh`.

(The reason that's a bug is that casting from `DataFile::Delegate*` to `c4Database*` requires subtracting an offset from the pointer value, because the former is not the first base class of the latter. But the compiler doesn't know to do that unless it's seen the declaration.)

Reply:
That definitely seems comment worthy